pkg/transport/queuestorage/file.go

```diff
@@ -1,164 +1,328 @@
+func (f *FileWriter) create(fileName string) error {
+	f.path = fileName
 	err := os.MkdirAll(filepath.Dir(f.path), os.ModePerm)
 	if err != nil {
-		log.Err(err).Msg("Unable to create all directories in file path")
+		log.Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Msg("unable to create all directories in file path")
 		return err
 	}
 
 	f.fd, err = os.OpenFile(f.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
 	if err != nil {
-		log.Err(err).Send()
+		log.Err(err).
+			Str("key", f.key).
+			Str("filePath", f.path).
+			Send()
 		return err
 	}
-	_ = f.fd.SetDeadline(f.expiry)
-
-	return nil
-}
 
-func (f *FileWriter) countDown() {
-	defer f.wg.Done()
-	defer f.ticker.Stop()
-
-	for {
-		select {
-		case now := <-f.ticker.C:
-			if now.After(f.expiry) {
-				log.Info().Str("key", f.key).
+	f.timer = time.AfterFunc(f.maxFileAge, func() {
+		f.mu.Lock()
+		defer f.mu.Unlock()
+		if !f.terminated {
+			log.Info().
+				Str("key", f.key).
+				Str("filePath", f.path).
+				Msg("maximum file age reached, closing file")
+
+			if err := f.close(false); err != nil {
+				log.Err(err).
+					Str("key", f.key).
 					Str("filePath", f.path).
-					Msg("Maximum file age reached, closing file")
-				err := f.Close()
-				if err != nil {
-					log.Err(err).Send()
-				}
-				return
+					Msg("ticker unable to close file")
 			}
 		}
-	}
+	})
+
+	return nil
 }
 
-func (f *FileWriter) Write(data []byte) (n int, err error) {
-	f.canWrite.Lock()
-	defer f.canWrite.Unlock()
+// ensureWritable checks for constraints before any write operation.
+func (f *FileWriter) ensureWritable(dataSize int64) (err error) {
+	var openNew bool
+	defer func() {
+		if openNew {
+			err = f.close(false)
+		}
+	}()
+
+	// check to see if we will hit our row limit
+	if f.maxRows <= 0 {
+		openNew = true
+		return
+	}
 
 	// check to see if we will hit our file size limit
-	fileInfo, err := f.fd.Stat()
+	var fileInfo os.FileInfo
+	fileInfo, err = f.fd.Stat()
 	if err != nil {
 		return
 	}
 
-	if fileInfo.Size()+int64(len(data)) > f.maxFileSize {
-		if err := f.Close(); err != nil {
-			log.Err(err).Send()
-		}
-		err = fmt.Errorf("file size limit reached")
+	newSize := fileInfo.Size() + dataSize
+	if newSize > f.maxFileSize {
+		openNew = true
 		return
 	}
 
+	return
+}
+
+// postOps uploads the current file and queues its detail
+func (f *FileWriter) postOps() error {
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("uploading file")
+
+	file, err := os.Open(f.path)
+	if err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to open file for upload")
+		return err
+	}
+
+	if err := f.storage.Upload(f.path, file); err != nil {
+		log.Error().Err(err).
+			Str("filePath", f.path).
+			Msg("unable to upload file")
+		return err
+	}
+
+	log.Info().
+		Str("key", f.key).
+		Str("filePath", f.path).
+		Msg("queuing file")
+
+	var bb []byte
+	if bb, err = json.Marshal(models.FileUploadMessage{
```

Comment:
I'd simplify this by moving the assignment out of the `if` statement, then there's no need for `var bb...`

Reply:
Scope.